### PR TITLE
Use !shouldfail rather than !mayfail on unit tests

### PR DESCRIPTION
**Pull request by thk123**
May fail runs the test and if it fails prints the output. Should fail runs the test and only prints the output if it actually passes. This makes it easier when looking at build output to determine which unit test has actually failed.

### CODE REVIEW

File: unit/util/irep_sharing.cpp [hunk 1]
```diff
@@ -7,7 +7,7 @@
 
 #ifdef SHARING
 
-SCENARIO("irept_sharing_trade_offs", "[!mayfail][core][utils][irept]")
+SCENARIO("irept_sharing_trade_offs", "[!shouldfail][core][utils][irept]")
```
Comment (tautschnig) [inline unit/util/irep_sharing.cpp:10]: I'm not sure I agree with this one: it should actually not fail, it's just a currently known bug in the implementation... (the same is true of the other change in this file). I'd suggest @NathanJPhillips comment on this.

Comment (thk123) [reply unit/util/irep_sharing.cpp:10]: (Nathan is AL atm, though this isn't critical so can wait if you don't agree with below)

I personally think if it starts _not_ failing, we should turn it on (i.e. remove the flag). This current flag allows it to fluctuate between working and not working without anyone ever noticing. By changing it to `!shouldfail`, if someone fixes it, the tests will fail and they will know to turn them on (and not accidentally write a bunch more!)

Comment (tautschnig) [reply unit/util/irep_sharing.cpp:10]: You might want to take a quick look at https://github.com/diffblue/cbmc/pull/1980 for the related discussion. I'm fine with your approach.

**Comment by antlechner** [comment]
I'm happy with this change (after finding it difficult to figure out which test was actually failing on CI), but could it also be documented that `[!shouldfail]` does not mean "this test should fail" at least in some cases? Either in a comment above every scenario line or somewhere central. Otherwise it might be hard to tell the difference at some point between known bugs and tests that we want to fail for some reason.

**Comment by thk123** [comment]
The travis output still seems to be outputing failing unit tests inspite of docs so I'll need to look into why that as otherwise this serves no purpose :upside_down_face:

**Comment by peterschrammel** [comment]
@thk123, why cannot we negate the assertion logic to make the tests actually pass (unless they are known bugs)?

**Comment by thk123** [comment]
@peterschrammel Well I think they are known bugs, so when they're fixed we just want to turn them on (rather than having to flip all the failing conditions).

**Review by NathanJPhillips — APPROVED** [review]

**Comment by owen-mc-diffblue** [comment]
I think the name "shouldfail" is unfortunate (I prefer "xfail", short for "expected to fail", which is what pytest uses). Still, I think the way that @thk123 is using it is good: in the long term no tests should be marked `[shouldfail]`; it should be used as a temporary solution when a test is failing and it can't be fixed immediately.

**Review by owen-mc-diffblue — APPROVED** [review]
Approved assuming it does what @thk123 originally thought it did in terms of reducing output and making it easier to see what is really failing.

**Comment by owen-mc-diffblue** [comment]
@thk123 I cherry-picked this commit and I also see that it still prints the failed assertions for `[!shouldfail]` tests. I don't see anything in the documentation that says it shouldn't do that (though I wish it did).

I still think that it is worth merging this PR so that if a bug gets inadvertently fixed we notice.

**Comment by thk123** [comment]
@owen-jones-diffblue yup you're right, must have misread it. I agree this is still better so I shall rebase and merge it (to ensure no shouldfail unit tests have been fixed in the meantime)

**Comment by owen-mc-diffblue** [comment]
And I guess you should check there aren't any new mayfail tests.

**Comment by thk123** [comment]
So JBMC unit tests are failing - I note that both the AWS build and the make builds *don't* fail which makes me think they are not running the JBMC unit tests. @peterschrammel are you aware of this?

**Comment by antlechner** [comment]
@thk123 @peterschrammel Has the AWS / make build / unit test issue been resolved?

**Comment by thk123** [comment]
Replaced by #5248

**Pull request closed without merging**